base/traits.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <type_traits>
+
+namespace principia {
+namespace base {
+namespace internal_traits {
+
+template<typename, typename = void, typename = void>
+struct has_templated_read_message : std::false_type {};
+
+template<typename T>
+struct has_templated_read_message<
+    T, std::void_t<decltype(&T::template ReadFromMessage<>)>>
+    : std::true_type {};
```

Comment:
`has_templatized_read_from_message`; but really we need the template parameters to all be defaulted, i.e., used for SFINAE; so perhaps `has_enabled_read_from_message`?

Reply:
Calling it `has_sfinae_read_from_message` because that's what it is.